File: Java/Inheritance/TypeCompatibility.java
//Program to study 
//1) Type Compatibility
//2) Late Binding
//3) Upcasting and Downcasting

//refer : Inheritance.txt
//see : TypeCompatibility.png

class A
{
 void f1()
 {
  System.out.println("A f1()");
 }

 void f2()
 {
  System.out.println("A f2()"); 
 }

}//A

class B extends A
{
 //override
 void f2(/* B this */) //B object back in B reference : Downcast
 {
  System.out.println("B f2() starts");
  f3();//extended code
  System.out.println("B f2() ends");

 }

 //extended code
 void f3()
 {
  System.out.println("B f3()"); 
 }

}//B

class TypeCompatibility
{
 void test(A refA )
 {
  System.out.println("---------------");
  refA.f1();
  refA.f2();//dynamic method dispatch or run time bound or late bound
  System.out.println("---------------");
 }

 public static void main(String args[])
 {
  A ref1 = new A(); //object of A
  B ref2 = new B(); //object of B

  TypeCompatibility tc = new TypeCompatibility();
  tc.test(ref1);
  tc.test(ref2);// B object in A reference : Upcast

 }//main
}//TypeCompatibility

File: c++/C++ Part2/inheritance/typeCompatibility.cpp
//Program to demonstrate : 
//Type Compatibility by Inheritance
//Runtime Binding
//refer : inheritance.txt
//see : typeCompatibility.png

#include <iostream>
using namespace std;

class A
{
public:
    void f1();
    virtual void f2();
};

void A :: f1()
{
    cout<<"\n A f1()";
}

void A :: f2()
{
    cout<<"\n A f2()";
}

class B : public A
{
public:
    void f2();
    void f3();
};

void B :: f2()
{
    cout<<"\n B f2() starts";
    f3();
    cout<<"\n B f2() ends";
}
void B :: f3()
{
    cout<<"\n B f3()";
}


void test(A * ptrA)
{
    cout<<"\n ----------";
    ptrA->f1();//inherited code
    ptrA->f2();//overridden code (runtime bound)
    //ptrA->f3();//extended code cannot be accessed
    cout<<"\n ----------";
}

int main()
{
    A objA;
    B objB;
    
    test(&objA);
    test(&objB);
    
   return 0;
}

